File: src/git/git.ts
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import path from 'node:path';

const exec = promisify(execFile);

/**
 * Git audit trail — every vault file operation gets a structured commit.
 * All paths are relative to vaultRoot.
 */
export class GitAudit {
  constructor(private vaultRoot: string) {}

  /**
   * Check if vault root is a git repo.
   * Runs: git rev-parse --git-dir
   */
  async isGitRepo(): Promise<boolean> {
    try {
      await exec('git', ['rev-parse', '--git-dir'], { cwd: this.vaultRoot });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Stage files and commit with structured message.
   * Message format: "ricket: {action} {details}"
   * Returns false if nothing to commit.
   */
  async commit(files: string[], message: string): Promise<boolean> {
    try {
      // Stage files
      if (files.length > 0) {
        await exec('git', ['add', ...files], { cwd: this.vaultRoot });
      }

      // Commit with --no-gpg-sign to avoid GPG issues
      await exec('git', ['commit', '-m', message, '--no-gpg-sign'], {
        cwd: this.vaultRoot,
      });

      return true;
    } catch (err) {
      const error = err as any;
      // Exit code 1 means nothing to commit
      if (error.code === 1 || error.status === 1) {
        return false;
      }
      // Other errors silently fail (best-effort audit)
      return false;
    }
  }

  /**
   * Stage a file move operation: git rm source, git add destination.
   * Then commit with message: "ricket: filed {source-basename} → {destination}"
   */
  async commitFileMove(source: string, destination: string): Promise<boolean> {
    try {
      const sourceBasename = path.basename(source);

      // Stage deletion and addition
      await exec('git', ['add', source], { cwd: this.vaultRoot });
      await exec('git', ['add', destination], { cwd: this.vaultRoot });

      // Commit with structured message
      const message = `ricket: filed ${sourceBasename} → ${destination}`;
      await exec('git', ['commit', '-m', message, '--no-gpg-sign'], {
        cwd: this.vaultRoot,
      });

      return true;
    } catch (err) {
      const error = err as any;
      // Exit code 1 means nothing to commit
      if (error.code === 1 || error.status === 1) {
        return false;
      }
      // Other errors silently fail (best-effort audit)
      return false;
    }
  }
}

File: internal/git/git.go
// Package git provides a structured git audit trail for vault operations.
package git

import (
	"fmt"
	"os/exec"
	"path/filepath"
)

// GitAudit commits vault file operations to git for audit trail.
type GitAudit struct {
	vaultRoot string
}

// New creates a GitAudit for the given vault root.
func New(vaultRoot string) *GitAudit {
	return &GitAudit{vaultRoot: vaultRoot}
}

// IsGitRepo returns true if the vault root is inside a git repository.
func (g *GitAudit) IsGitRepo() bool {
	cmd := exec.Command("git", "rev-parse", "--git-dir")
	cmd.Dir = g.vaultRoot
	return cmd.Run() == nil
}

// Commit stages the given files and commits with message.
// Uses --no-gpg-sign to avoid GPG issues.
// Returns false if nothing to commit or on any error (best-effort audit).
func (g *GitAudit) Commit(files []string, message string) bool {
	if len(files) > 0 {
		args := append([]string{"add"}, files...)
		cmd := exec.Command("git", args...)
		cmd.Dir = g.vaultRoot
		if err := cmd.Run(); err != nil {
			return false
		}
	}

	cmd := exec.Command("git", "commit", "-m", message, "--no-gpg-sign")
	cmd.Dir = g.vaultRoot
	if err := cmd.Run(); err != nil {
		return false
	}
	return true
}

// CommitFileMove stages source (deleted) and destination (added), then commits.
// Message format: "ricket: filed {source-basename} → {destination}"
func (g *GitAudit) CommitFileMove(source, destination string) bool {
	sourceBasename := filepath.Base(source)

	addSrc := exec.Command("git", "add", source)
	addSrc.Dir = g.vaultRoot
	addSrc.Run() //nolint:errcheck

	addDst := exec.Command("git", "add", destination)
	addDst.Dir = g.vaultRoot
	addDst.Run() //nolint:errcheck

	message := fmt.Sprintf("ricket: filed %s → %s", sourceBasename, destination)
	cmd := exec.Command("git", "commit", "-m", message, "--no-gpg-sign")
	cmd.Dir = g.vaultRoot
	return cmd.Run() == nil
}
